File: W3/homeWork_W3/Java/PenPaper/Paper.java
public class Paper {
    private StringBuffer content;
    private int maxSymbols;
    private int symbols;
    
    public Paper(int maxSymbols) {
        this.content = new StringBuffer();
        this.maxSymbols = maxSymbols;
        this.symbols = 0;
    }

    public Paper() {
        this(4096);
    }

    public final int getMaxSymbols() {
        return this.maxSymbols;
    }
    
    public final int getSymbols() {
        return this.symbols;
    }
    
    public void addContent(String message) throws OutOfSpaceException {
        this.symbols = content.length() + message.length();
        
        if ( this.maxSymbols <= 0 ) {
            throw new OutOfSpaceException();
            }
        if ( content.length() == this.maxSymbols ) {
            throw new OutOfSpaceException();
        }
        if ( this.symbols > this.maxSymbols ) {
            content.append(message.substring(0, this.maxSymbols - content.length()));
            return;
        }
        content.append(message);
    }
    
    public void show() {
        System.out.println(content);
    }
}

File: W3/homeWork_W3/С++/Pen&Paper/Paper.cpp
#include <string>
#include "Paper.h"

using namespace std;

Paper::Paper(int maxSymbols) {
    this->symbols = 0;
    this->maxSymbols = maxSymbols;    
}

Paper::~Paper() {
    
}

int Paper::getMaxSymbols() const {
    return this->maxSymbols;
}

int Paper::getSymbols() const {
    return this->symbols;
}

void Paper::addContent(const string& message) {
    int start = 0;
    
    if ( this->maxSymbols <= 0 ) {
        throw OutOfSpace();
    }
    string newMessage = message.substr(start, this->maxSymbols);
    this->maxSymbols -= newMessage.length();
    
    this->content += newMessage;
    this->symbols = this->content.length();
}

void Paper::show() const {
    cout << this->content << endl;
}
